docs/design-proposals/2022-04 TraceQL Concepts.md

```diff
@@ -0,0 +1,198 @@
+`{ http.status >= 200 && http.status < 300 }`
+
+Note that the second expression requires both conditions to be true on the same span. The entire expression inside of `{}` must be evaluated as true on a single span for it to be included in the resultset.
+
+## Combining Spansets
+
+Logical operators combine sets of spans. For example, if we wanted to find a trace that went through two specific regions:  
+`{ region = "eu-west-0" } && { region = "eu-west-1" } `
+
+Note the difference between the above and the following:  
+`{ region = "eu-west-0"  && region = "eu-west-1" } `
+
+The second expression will return no traces because it's impossible for both conditions to be simultaneously true on the same span.
+
+### Structural operators
+
+Structural operators evaluate traces based on the span relationships within the span tree. 
+
+`{ } >> { }`  
+  This is the descendant operator. The spans returned from this operator will match the right hand side conditions while also being descendants of spans that match the left hand side conditions.
+
+`{ } > { }`  
+  This is the child operator. The spans returned from this operator will match the right hand side conditions while also being children of spans that match the left hand side conditions.
+
+`{ } ~ { }`  
+  This is the sibling operator. The spans returned from this operator will match the right hand side conditions while also being siblings of spans that match the left hand side conditions.
+
+## Aggregators
+
+All of the above expressions involve asking questions about individual spans. However, sometimes we want to ask questions about a set of spans. For that we have aggregate functions.
+
+**Examples**
+
+Find traces where the total number of spans is greater than 10:  
+`count() > 10`
+
+Find traces where the average duration is greater than 1s:  
+`avg(duration) > 1s`
+
+## Expression Pipelining
+
+Pipelining allows us to "pipe" a set of spans from one expression to the next. This is particularly useful if you want to perform an aggregate over a subset of a trace.
+
+**Examples**
+
+Find traces where that have more than 3 spans with an attribute`http.status` with a value of `200`:  
+`{ http.status = 200 } | count() > 3`
+
+## Grouping
+
+Grouping allows us to take a trace and break it down into sets of spans that are evaluated by future pipeline entries. Each set of spans created by the group is _individually_ evaluated by downstream expressions.
+
+**Examples**
+
+Find traces that have more than 5 spans in any region:  
+`by(region) | count() > 5 `
+
+## Examples
+
+Any span matches an attribute:  
+`{ namespace = "prod" }`
+
+Two attributes appear on the same span:  
+`{ namespace = "prod" && http.status = 200 }`
+
+Two attributes appear anywhere within a trace:  
+`{ namespace = "prod" } && { http.status = 200 }`
+
+Any span has a duration over one second:  
+`{ duration > 1s }`
+
+The trace as a whole has a duration of over one second:  
+`max(end) - min(start) > 1s`
```

Comment:
I expect Tempo to store the final trace duration determined exactly the same way, so could this simplified?  Thinking that this is an inefficient precedence to set, as it requires scanning all spans for two values instead of the single value per trace.
```suggestion
`duration > 1s`
```

I realize this kind of conflicts with the span-level duration, so maybe another intrinsic property like 'totalDuration`.

Reply:
There is currently no way to reference the "trace" as a thing in traceql. All of the operations are selecting and filtering sets of spans. There are a few "trace level" attributes that I wonder if it would make sense to include or not, but we would have to think quite carefully about we would integrate it into our current "pipeline of spansets" approach.

Examples of trace level attributes
- duration
- rootSpanName
- ??

To help with the second one we have considered adding a "root" intrinsic so you could ask something like:

`{ name = "do the thing" && root = true }`

but that's still up in the air.

Reply:
Trace duration is absolutely an attribute that people would query often. Perhaps having a `root` level would be already sufficient in terms of performance? The trace duration would then be `max(end) - root.start`?